File: src/internal/client/scans.go
package client

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
)

// Fields that the resource schema gives a Default must NOT be tagged
// omitempty. The schema promises the attribute always has a value, so the plan
// always carries one; dropping it from the wire when it happens to be the zero
// value means the API never learns the user cleared it, echoes the stale value
// back, and the apply fails on a mismatch the provider itself created.
// omitempty stays only where absence is genuinely meaningful.
type ScanSettings struct {
	Name           string   `json:"name"`
	Description    string   `json:"description"`
	PolicyID       int      `json:"policy_id,omitempty"`
	FolderID       int      `json:"folder_id,omitempty"`
	ScannerID      int      `json:"scanner_id,omitempty"`
	TextTargets    string   `json:"text_targets,omitempty"`
	TagTargets     []string `json:"tag_targets,omitempty"`
	FileTargets    string   `json:"file_targets,omitempty"`
	Launch         string   `json:"launch,omitempty"`
	Enabled        bool     `json:"enabled"`
	Starttime      string   `json:"starttime,omitempty"`
	RRules         string   `json:"rrules,omitempty"`
	Timezone       string   `json:"timezone,omitempty"`
	Emails         string   `json:"emails,omitempty"`
	ScanTimeWindow int      `json:"scan_time_window,omitempty"`
}

type ScanCreateRequest struct {
	UUID     string       `json:"uuid"`
	Settings ScanSettings `json:"settings"`
}

type ScanUpdateRequest struct {
	UUID     string       `json:"uuid,omitempty"`
	Settings ScanSettings `json:"settings"`
}

type ScanCreateResponse struct {
	Scan ScanDetail `json:"scan"`
}

// ScanUpdateResponse carries whatever PUT /scans/{id} answered with.
//
// Three shapes are in play and the provider has to survive all of them: the
// documented response is the scan object *bare* -- not wrapped in "scan" the
// way POST /scans wraps it -- while tenants have been seen to answer with no
// body at all. Scan is nil for an empty body, which means "the update was not
// echoed", not "the update echoed empty values".
type ScanUpdateResponse struct {
	Scan *ScanDetail
}

func (r *ScanUpdateResponse) UnmarshalJSON(data []byte) error {
	if string(data) == "null" {
		r.Scan = nil
		return nil
	}

	var wrapped struct {
		Scan *ScanDetail `json:"scan"`
	}
	if err := json.Unmarshal(data, &wrapped); err == nil && wrapped.Scan != nil {
		r.Scan = wrapped.Scan
		return nil
	}

	var bare ScanDetail
	if err := json.Unmarshal(data, &bare); err != nil {
		return err
	}
	r.Scan = &bare
	return nil
}

// UnmarshalJSON decodes a scan object and records which keys it carried.
func (d *ScanDetail) UnmarshalJSON(data []byte) error {
	type plain ScanDetail
	var decoded plain
	if err := json.Unmarshal(data, &decoded); err != nil {
		return err
	}
	keys, err := presentKeys(data)
	if err != nil {
		return err
	}
	*d = ScanDetail(decoded)
	d.PresentKeys = keys
	return nil
}

// UnmarshalJSON decodes an info object and records which keys it carried.
func (i *ScanInfo) UnmarshalJSON(data []byte) error {
	type plain ScanInfo
	var decoded plain
	if err := json.Unmarshal(data, &decoded); err != nil {
		return err
	}
	keys, err := presentKeys(data)
	if err != nil {
		return err
	}
	*i = ScanInfo(decoded)
	i.PresentKeys = keys
	return nil
}

// presentKeys returns the top-level keys of a JSON object, sorted.
func presentKeys(data []byte) ([]string, error) {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}
	keys := make([]string, 0, len(raw))
	for key := range raw {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys, nil
}

type ScanDetail struct {
	// PresentKeys lists the keys the response actually carried, sorted. Key
	// names only, never values, so it is safe to quote in a diagnostic: when a
	// field comes back empty it answers "did the API omit it, or is the whole
	// object thinner than it should be?" -- which is the difference between a
	// normalising tenant and something other than Tenable.io answering.
	PresentKeys []string `json:"-"`

	ID   int    `json:"id"`
	UUID string `json:"uuid"`
	Name string `json:"name"`
	// Description is a pointer so that "the response did not carry the field"
	// stays distinguishable from "the response carried an empty string". Only
	// the second can be compared with configuration; treating the first as ""
	// is what made a tenant that omits the key look like one that wiped the
	// text. See ScanInfo.Description.
	Description          *string `json:"description"`
	PolicyID             int     `json:"policy_id"`
	FolderID             int     `json:"folder_id"`
	ScannerID            int     `json:"scanner_id"`
	TextTargets          string  `json:"text_targets"`
	Starttime            string  `json:"starttime"`
	RRules               string  `json:"rrules"`
	Timezone             string  `json:"timezone"`
	Emails               string  `json:"emails"`
	Enabled              bool    `json:"enabled"`
	Launch               string  `json:"launch"`
	ScanTimeWindow       int     `json:"scan_time_window"`
	Status               string  `json:"status"`
	CreationDate         int     `json:"creation_date"`
	LastModificationDate int     `json:"last_modification_date"`
	Type                 string  `json:"type"`
}

type ScanDetailsResponse struct {
	Info ScanInfo `json:"info"`
}

type ScanInfo struct {
	// PresentKeys is ScanDetail.PresentKeys for the details response.
	PresentKeys []string `json:"-"`

	ID   int    `json:"object_id"`
	UUID string `json:"uuid"`
	Name string `json:"name"`
	// Description is absent from the documented info schema and from live
	// responses: GET /scans/{id} reports a scan *result*, not the settings that
	// were submitted. nil therefore means "unknown", and the provider must not
	// compare it with configuration or write it over state -- doing so reported
	// every described scan as having had its description wiped.
	Description          *string `json:"description"`
	PolicyID             int     `json:"policy_id"`
	FolderID             int     `json:"folder_id"`
	ScannerID            int     `json:"scanner_id"`
	Targets              string  `json:"targets"`
	Starttime            string  `json:"starttime"`
	RRules               string  `json:"rrules"`
	Timezone             string  `json:"timezone"`
	Emails               string  `json:"notification_email_address"`
	Enabled              bool    `json:"enabled"`
	Launch               string  `json:"launch"`
	ScanTimeWindow       int     `json:"scan_time_window"`
	Status               string  `json:"status"`
	CreationDate         int     `json:"creation_date"`
	LastModificationDate int     `json:"last_modification_date"`
	ScanType             string  `json:"scan_type"`
	TemplateUUID         string  `json:"scanner_name"`
}

type ScansListResponse struct {
	Scans []ScanListItem `json:"scans"`
}

type ScanListItem struct {
	ID                   int    `json:"id"`
	UUID                 string `json:"uuid"`
	Name                 string `json:"name"`
	Description          string `json:"description"`
	FolderID             int    `json:"folder_id"`
	Type                 string `json:"type"`
	Status               string `json:"status"`
	Enabled              bool   `json:"enabled"`
	CreationDate         int    `json:"creation_date"`
	LastModificationDate int    `json:"last_modification_date"`
}

func (c *Client) CreateScan(ctx context.Context, req ScanCreateRequest) (*ScanCreateResponse, error) {
	var resp ScanCreateResponse
	if err := c.Post(ctx, "/scans", req, &resp); err != nil {
		return nil, fmt.Errorf("creating scan: %w", err)
	}
	return &resp, nil
}

func (c *Client) GetScan(ctx context.Context, scanID int) (*ScanDetailsResponse, error) {
	var resp ScanDetailsResponse
	if err := c.Get(ctx, fmt.Sprintf("/scans/%d", scanID), &resp); err != nil {
		return nil, fmt.Errorf("getting scan: %w", err)
	}
	return &resp, nil
}

// UpdateScan returns the PUT echo when there is one. The response is the only
// place an update's description comes back from, since the scan details
// endpoint does not report it; resp.Scan is nil when the tenant answers with an
// empty body.
func (c *Client) UpdateScan(ctx context.Context, scanID int, req ScanUpdateRequest) (*ScanUpdateResponse, error) {
	var resp ScanUpdateResponse
	if err := c.Put(ctx, fmt.Sprintf("/scans/%d", scanID), req, &resp); err != nil {
		return nil, fmt.Errorf("updating scan: %w", err)
	}
	return &resp, nil
}

func (c *Client) DeleteScan(ctx context.Context, scanID int) error {
	if err := c.Delete(ctx, fmt.Sprintf("/scans/%d", scanID)); err != nil {
		return fmt.Errorf("deleting scan: %w", err)
	}
	return nil
}

func (c *Client) ListScans(ctx context.Context, folderID *int) (*ScansListResponse, error) {
	path := "/scans"
	if folderID != nil {
		path = fmt.Sprintf("/scans?folder_id=%d", *folderID)
	}
	var resp ScansListResponse
	if err := c.Get(ctx, path, &resp); err != nil {
		return nil, fmt.Errorf("listing scans: %w", err)
	}
	return &resp, nil
}

File: mockapi/tenableio_mock/routers/scans.py
"""Scans.

The read and write shapes diverge more here than anywhere else outside tags,
and the divergence is faithful:

* ``POST /scans`` returns the new scan wrapped as ``{"scan": {...}}``.
* ``GET /scans/{id}`` returns ``{"info": {...}}`` -- a *different* object whose
  keys are renamed: the identifier is ``object_id`` rather than ``id``, the
  target list is ``targets`` rather than ``text_targets``, the recipient list is
  ``notification_email_address`` rather than ``emails``, and the template is
  reported under ``scanner_name``.
* ``GET /scans/{id}`` does **not** report ``description`` at all. It describes a
  scan result rather than the settings that were submitted, and the documented
  ``info`` schema has no such field. The mock used to include one; that single
  courtesy hid a production bug, because a provider can read state back from
  this endpoint, find ``""``, and conclude the description was wiped. Turn
  ``MOCK_SCAN_DETAILS_DESCRIPTION`` on to get the forgiving shape back.
* ``GET /scans`` returns a slimmer list item again.
* ``PUT /scans/{id}`` returns an empty body by default. With
  ``MOCK_SCAN_UPDATE_ECHO=object`` it returns the scan object *bare* -- the
  documented shape, and notably not wrapped in ``"scan"`` the way the create
  response is. That echo is the only place an update's description comes back
  from, so a provider has to verify what it wrote when the body is there and
  accept not being able to when it is not.

Anything that flattens these into one shape is not reproducing the API.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request, Response
from fastapi.responses import JSONResponse

from ..config import ScanUpdateEcho
from ..errors import bad_request, not_found
from ..models import ScanCreate, ScanSettings, ScanUpdate
from ..store import Store, unix_seconds
from ._common import get_settings, get_store, reject_unknown, resolve_on_omit

router = APIRouter(tags=["scans"])

MAX_DESCRIPTION = 3000
LAUNCH_VALUES = {"ON_DEMAND", "DAILY", "WEEKLY", "MONTHLY", "YEARLY"}


def _validate(settings_body: ScanSettings, store: Store) -> None:
    if not settings_body.name or not settings_body.name.strip():
        raise bad_request("The scan name is required.")
    if settings_body.description is not None and len(settings_body.description) > MAX_DESCRIPTION:
        raise bad_request(
            f"The scan description exceeds the maximum length of "
            f"{MAX_DESCRIPTION} characters."
        )
    if settings_body.launch is not None and settings_body.launch not in LAUNCH_VALUES:
        raise bad_request(
            f"launch must be one of: {', '.join(sorted(LAUNCH_VALUES))}."
        )
    if settings_body.folder_id and settings_body.folder_id not in store.folders:
        raise bad_request(
            f"The specified folder {settings_body.folder_id} does not exist.",
            code="not_found",
        )
    if settings_body.scanner_id and settings_body.scanner_id not in store.scanners:
        raise bad_request(
            f"The specified scanner {settings_body.scanner_id} does not exist.",
            code="not_found",
        )
    if settings_body.policy_id and settings_body.policy_id not in store.policies:
        raise bad_request(
            f"The specified policy {settings_body.policy_id} does not exist.",
            code="not_found",
        )


def _find(store: Store, scan_id: int) -> dict[str, Any]:
    scan = store.scans.get(scan_id)
    if scan is None:
        raise not_found(f"Scan {scan_id} was not found.")
    return scan


def _detail_payload(scan: dict[str, Any]) -> dict[str, Any]:
    """The shape ``POST /scans`` wraps in ``{"scan": ...}``."""
    return {
        "id": scan["id"],
        "uuid": scan["uuid"],
        "name": scan["name"],
        "description": scan["description"],
        "policy_id": scan["policy_id"],
        "folder_id": scan["folder_id"],
        "scanner_id": scan["scanner_id"],
        "text_targets": scan["text_targets"],
        "starttime": scan["starttime"],
        "rrules": scan["rrules"],
        "timezone": scan["timezone"],
        "emails": scan["emails"],
        "enabled": scan["enabled"],
        "launch": scan["launch"],
        "scan_time_window": scan["scan_time_window"],
        "status": scan["status"],
        "creation_date": scan["creation_date"],
        "last_modification_date": scan["last_modification_date"],
        "type": scan["type"],
    }


def _info_payload(scan: dict[str, Any], *, describe: bool = False) -> dict[str, Any]:
    """The renamed shape ``GET /scans/{id}`` wraps in ``{"info": ...}``.

    ``describe`` adds the ``description`` key that live Tenable.io leaves out.
    """
    info: dict[str, Any] = {
        "object_id": scan["id"],
        "uuid": scan["uuid"],
        "name": scan["name"],
        "policy_id": scan["policy_id"],
        "folder_id": scan["folder_id"],
        "scanner_id": scan["scanner_id"],
        "targets": scan["text_targets"],
        "starttime": scan["starttime"],
        "rrules": scan["rrules"],
        "timezone": scan["timezone"],
        "notification_email_address": scan["emails"],
        "enabled": scan["enabled"],
        "launch": scan["launch"],
        "scan_time_window": scan["scan_time_window"],
        "status": scan["status"],
        "creation_date": scan["creation_date"],
        "last_modification_date": scan["last_modification_date"],
        "scan_type": scan["type"],
        "scanner_name": scan["template_uuid"],
    }
    if describe:
        info["description"] = scan["description"]
    return info


def _list_payload(scan: dict[str, Any]) -> dict[str, Any]:
    return {
        "id": scan["id"],
        "uuid": scan["uuid"],
        "name": scan["name"],
        "description": scan["description"],
        "folder_id": scan["folder_id"],
        "type": scan["type"],
        "status": scan["status"],
        "enabled": scan["enabled"],
        "creation_date": scan["creation_date"],
        "last_modification_date": scan["last_modification_date"],
    }


@router.post("/scans")
async def create_scan(body: ScanCreate, request: Request) -> dict[str, Any]:
    store, settings = get_store(request), get_settings(request)
    reject_unknown(body, settings)
    if not body.uuid or not body.uuid.strip():
        raise bad_request("The template uuid is required.")

    with store.lock:
        _validate(body.settings, store)
        stamp = unix_seconds(store.now())
        s = body.settings
        scan = {
            "id": store.next_id(),
            "uuid": store.next_uuid(),
            "name": s.name,
            "description": s.description or "",
            "policy_id": s.policy_id or 0,
            "folder_id": s.folder_id or 0,
            "scanner_id": s.scanner_id or 0,
            "text_targets": s.text_targets or "",
            "tag_targets": s.tag_targets or [],
            "starttime": s.starttime or "",
            "rrules": s.rrules or "",
            "timezone": s.timezone or "",
            "emails": s.emails or "",
            "enabled": s.enabled,
            "launch": s.launch or "ON_DEMAND",
            "scan_time_window": s.scan_time_window or 0,
            "status": "empty",
            "creation_date": stamp,
            "last_modification_date": stamp,
            "type": "public",
            "template_uuid": body.uuid,
        }
        store.scans[scan["id"]] = scan
        return {"scan": _detail_payload(scan)}


@router.get("/scans")
async def list_scans(request: Request, folder_id: int | None = None) -> dict[str, Any]:
    store = get_store(request)
    with store.lock:
        scans = list(store.scans.values())
        if folder_id is not None:
            scans = [s for s in scans if s["folder_id"] == folder_id]
        return {"scans": [_list_payload(s) for s in scans]}


@router.get("/scans/{scan_id}")
async def get_scan(scan_id: int, request: Request) -> dict[str, Any]:
    store, settings = get_store(request), get_settings(request)
    with store.lock:
        return {
            "info": _info_payload(
                _find(store, scan_id),
                describe=settings.quirks.scan_details_description,
            )
        }


@router.put("/scans/{scan_id}")
async def update_scan(scan_id: int, body: ScanUpdate, request: Request) -> Response:
    store, settings = get_store(request), get_settings(request)
    reject_unknown(body, settings)

    with store.lock:
        scan = _find(store, scan_id)
        _validate(body.settings, store)
        s = body.settings

        scan["name"] = s.name
        scan["description"] = resolve_on_omit(
            s, "description", scan["description"], settings.quirks.on_omitted_description
        )
        scan["enabled"] = s.enabled
        for field, key in (
            ("policy_id", "policy_id"),
            ("folder_id", "folder_id"),
            ("scanner_id", "scanner_id"),
            ("text_targets", "text_targets"),
            ("tag_targets", "tag_targets"),
            ("starttime", "starttime"),
            ("rrules", "rrules"),
            ("timezone", "timezone"),
            ("emails", "emails"),
            ("launch", "launch"),
            ("scan_time_window", "scan_time_window"),
        ):
            value = getattr(s, field)
            if value is not None:
                scan[key] = value
        if body.uuid:
            scan["template_uuid"] = body.uuid
        scan["last_modification_date"] = unix_seconds(store.now())

        if settings.quirks.scan_update_echo is ScanUpdateEcho.OBJECT:
            # Bare, not wrapped in "scan": the update response and the create
            # response disagree about that in the docs, and reproducing only the
            # create shape would let a provider hard-code the wrong one.
            return JSONResponse(_detail_payload(scan))
    return Response(status_code=200)


@router.delete("/scans/{scan_id}")
async def delete_scan(scan_id: int, request: Request) -> Response:
    store = get_store(request)
    with store.lock:
        _find(store, scan_id)
        del store.scans[scan_id]
    return Response(status_code=200)
